_ja/tutorials/scala-for-java-programmers.md

```diff
@@ -0,0 +1,559 @@
+  一方で、新しい処理を追加するのは、独立した関数を定義するだけなので簡単です。
+
+パターンマッチについてさらに知るために、演算式に対する別の処理、数式微分を定義してみましょう。
+この処理についての次のような規則をご存知かもしれません。
+
+1. 和の微分係数は、微分係数の和です。
+2. 何らかの変数 `v` の微分係数は、変数 `v` に対して微分されるなら1、そうでなければ0です。
+3. 定数の微分係数は0です。
+
+これら規則はほぼ文字通り Scala コードに翻訳でき、次の定義が得られます。
+
+    def derive(t: Tree, v: String): Tree = t match {
+      case Sum(l, r) => Sum(derive(l, v), derive(r, v))
+      case Var(n) if (v == n) => Const(1)
+      case _ => Const(0)
+    }
+
+この関数はパターンマッチに関連して2つの新しい概念を紹介しています。
+はじめに、変数への `case` 式が、`if` キーワードに続く式、*ガード*を持っています。
+このガードは、その式が真でない限りパターンマッチが成功するのを防ぎます。
+ここでのガードの使用は、微分される変数の名前が微分の変数 `v` と同じときにだけ、定数`1` を返すことを保証するためで。
+ここで使われているパターンマッチの新しい特徴2つ目は、`_` で書かれている*ワイルドカード*です。
+それはどんな値にもマッチして、名前をつけないパターンです。
+
+パターンマッチの全力をまだ探っていませんが、この文章を短くするためにここで止めておきます。
+実例について上の2つの関数がどのように実行するかをまだ見たいですね。
+そのために、式 `(x+x)+(7+y)` にいくつかの処理を実行する簡単な `main` 関数を書いてみましょう。
+まず環境における値 `{ x -> 5, y -> 7 }` を計算し、次に `x` そして `y` についての微分係数を計算します。
+
+    def main(args: Array[String]) {
+      val exp: Tree = Sum(Sum(Var("x"),Var("x")),Sum(Const(7),Var("y")))
+      val env: Environment = { case "x" => 5 case "y" => 7 }
+      println("Expression: " + exp)
+      println("Evaluation with x=5, y=7: " + eval(exp, env))
+      println("Derivative relative to x:\n " + derive(exp, "x"))
+      println("Derivative relative to y:\n " + derive(exp, "y"))
+    }
+
+コンパイルする前に、`Environment` 型、`eval`、`derive` そして `main` メソッドを `Calc` オブジェクトで包む必要があります。
+このプログラムを実行することで、期待する結果が得られます。
+
+    Expression: Sum(Sum(Var(x),Var(x)),Sum(Const(7),Var(y)))
+    Evaluation with x=5, y=7: 24
+    Derivative relative to x:
+     Sum(Sum(Const(1),Const(1)),Sum(Const(0),Const(0)))
+    Derivative relative to y:
+     Sum(Sum(Const(0),Const(0)),Sum(Const(0),Const(1)))
+
+出力を観察することで、微分係数の結果はユーザーに示されるより前にシンプルにすべきことに気づきます。
+パターンマッチを使った基本的なシンプル化の定義は興味深い（ですが驚くほど一筋縄ではいかない）問題なので、読者への練習問題として残しておきます。
+
+## トレイト
+
+スーパークラスからコードを継承するのとは別に、Scala クラスは1つ以上の*トレイト*からコードを取り込めます。
+
+おそらく Java プログラマーにとってトレイトを理解するもっとも簡単な方法は、コードを含むことができるインターフェースとしてとらえることでしょう
+Scala では、クラスがトレイトから継承するとき、トレイトのインターフェースを実装し、トレイトに含まれるコードをすべて継承します。
+
+（注：Java 8 からは、Java インターフェースも同様にコードを含めるようになりました。`default` キーワードをつけるか、または static メソッドとして)
+
+トレイトの有用性を見るため、古典的な例、順序付きオブジェクトを見てみましょう。
+あるクラスのオブジェクト同士を順序比較できると、例えば並び替え（ソート）できて、便利なことが多いです。
+Java では、比較できるオブジェクトは `Comparable` インターフェースを実装します。
+Scala では、`Comparable` 相当を `Ord` という名前のトレイトとして定義することで、Java よりも少しだけ良くなっています。
+
+オブジェクトを比較するとき、6つの異なる述語（より小さい、より小さいか等しい、等しい、等しくない、より大きいか等しい、より大きい）が役に立ちます。
+しかし、これらすべてを定義するのは、これら6つのうち4つは残りの2つを使って表現できるので、退屈です。
+つまり、例えば「等しい」と「より小さい」の述語があれば、他を表現できるということです。
+Scala では、これらの観察結果は以下のトレイト宣言にうまくとらえられます。
+
+    trait Ord {
+      def < (that: Any): Boolean
+      def <=(that: Any): Boolean =  (this < that) || (this == that)
+      def > (that: Any): Boolean = !(this <= that)
+      def >=(that: Any): Boolean = !(this < that)
+    }
+
+この定義は、Java の `Comparable` インターフェースと同じ働きをする `Ord` という新しい型と、3つの述語のデフォルト実装を4つ目の抽象述語を使うことで作成しています。
+相等と不等の述語はすべてのオブジェクトにデフォルトで存在するので現れていません。
```

Comment:
原文: The predicates for equality and inequality do not appear here since they are by default present in all objects.
訳案: 等式と不等式の述語はすべてのオブジェクトにデフォルトで存在するので、ここには現れていません。

相等というのが聞き慣れないので、equality の訳語としては等式とかの方が分かりやすいかなと思いました。

Reply:
そうですね。
`(in)equality` はどう訳すか迷ったのですが、等式・不等式なら聞き慣れてるので、分かりやすいですね。
https://github.com/scala/docs.scala-lang/pull/1416/commits/eb999c85055ecb51c3481a48d3b7942dd2be8213 で修正しました